### ISSUE TITLE
Prometheus receiver may not be honouring the memory limiting and seems to have a memory leak

### ISSUE THREAD

**Issue by gillg** [open-telemetry/opentelemetry-collector-contrib#4916]
**Describe the bug**
As discussed here https://cloud-native.slack.com/archives/C01N6P7KR6W/p1621329415004800 the memory limiter drop lot of scrapped datapoints after some times (minutes/hours) runing.
@tigrannajaryan suspects that prometheus receiver doesn't honor the memory limit.

After some minutes I have lot of errors like "Memory usage is above soft limit. Dropping data." if my scraped targets have a lot of metrics.

I also suspect something related to some "high cardinality" metrics.
I have some metrics like : 
`windows_process_cpu_time_total{creating_process_id="102916",mode="privileged",otel_job="windows-exporter",process="Agent",process_id="141812"} 0.640625`
These metrics are already filtered by process name, but the resulting metrics can be spread accross multiple "process_id" because this process can die and live multiple times...

**Steps to reproduce**
I don't know exactly but scraping 2 targets with 40 and 3257 timeseries and the following config I have issues.
Scraped targets metrics
```
scrape_samples_scraped{otel_job="otel-collector"} 40
scrape_samples_scraped{otel_job="windows-exporter"} 3257
```
Config :
```yaml
receivers:
  prometheus:
    config:
      scrape_configs:
        - job_name: 'otel-collector'
          scrape_interval: 10s
          static_configs:
            - targets: ['localhost:8888']
          relabel_configs:
          # Trick because otel collector not expose the job
          - action: replace
            replacement: otel-collector
            target_label: otel_job
        - job_name: windows-exporter
          scrape_interval: 10s
          static_configs:
            - targets: ['localhost:9182']
          relabel_configs:
          # Trick because otel collector not expose the job
          - action: replace
            replacement: windows-exporter
            target_label: otel_job

exporters:
  logging:
    loglevel: debug
  prometheus:
    endpoint: "0.0.0.0:9095"
    send_timestamps: false
    metric_expiration: 60m

processors:
  batch:
    send_batch_size: 50
    timeout: 5s
  memory_limiter:
    ballast_size_mib: 128
    check_interval: 1s
    limit_mib: 256
    spike_limit_percentage: 30

extensions:
  health_check:

service:
  extensions: [health_check]
  pipelines:
    metrics:
      receivers: [prometheus]
      processors: [memory_limiter, batch]
      exporters: [prometheus]
```

**What did you expect to see?**
A clear and concise description of what you expected to see.

**What did you see instead?**
A clear and concise description of what you saw instead.

**What version did you use?**
Custom Version: https://github.com/open-telemetry/opentelemetry-collector/pull/3116/commits/bf36d80c510a5f3bd38a0d9e373f7dd14167a780

**Environment**
OS: Windows
Compiler(if manually compiled): Go 1.16

**Additional context**
After some minutes some metrics start to be droped with lot of errors "Memory usage is above soft limit. Dropping data."
After some hours, memory can be above than 380Mb

**Comment by locmai** [comment]
We are having the same issue with Windows hosts as well on our production. Do we have some updates on this one?

**Comment by locmai** [comment]
Our case is the otel collector going up to 26GB for one instance, and the rate is around 1GB per day.

**Comment by locmai** [comment]
So I have followed this guide https://blog.detectify.com/2019/09/05/how-we-tracked-down-a-memory-leak-in-one-of-our-go-microservices/ 
After doing all the chasing after the heap profile, I see that the memory allocated for otel is not even peak the memory allocated for the service.

And the final solution of the article works for me: Set the `GODEBUG=madvdontneed=1` env var for the runtime and restart the service.

**Comment by gillg** [comment]
@locmai can you confirm but this problem seems having disapear in recent otel collector contrib versions.
I was making an auto restart once or twice a day before but it seems not useful anymore.

**Comment by locmai** [comment]
I haven't tried the latest one since then yet, I figured out that the problem was that the ./metrics endpoint is not being called (if I called it via Prometheus or some other clients) it would drop the memory. (Just on those Windows boxes btw)
So my workarounds were:
- Either auto restart like you did
- Make the call to the endpoint probably (I set the Prometheus endpoint incorrectly which led to the endpoint never got called).
- Have a cronjob to call the endpoint as well.

---
Our patching routine will start soon and we will upgrade to the latest one, I will observe and update on this issue.
